wallet/src/Pos/Wallet/Web/Tracking/Restore.hs

```diff
@@ -0,0 +1,103 @@
+                 , MonadMask m
+                 , MonadSlotsData ctx m
+                 , MonadUnliftIO m
+                 ) => WalletDecrCredentials -> m ()
+restoreWallet credentials = do
+    db <- askWalletDB
+    let (_, walletId) = credentials
+    modifyLoggerName (const "syncWalletWorker") $ do
+        logInfo "New Restoration request for a wallet..."
+        genesisBlockHeaderE <- firstGenesisHeader
+        case genesisBlockHeaderE of
+            Left syncError -> processSyncError syncError
+            Right genesisBlock -> do
+                restoreGenesisAddresses db credentials
+                restoreWalletBalance db credentials
+                -- At this point, we consider ourselves synced with the UTXO up-to the
+                -- 'RestorationBlockDepth' we compute now. During 'syncWalletWithBlockchain',
+                -- we will restore the wallet history from the beginning of the chain by ignoring
+                -- any Utxo changes, but we will always add transactions to the pool of known ones.
+                -- By doing so, the BListener is free to track new blocks (both in terms of balance update
+                -- & tx tracking), allowing the user to use the wallet even if is technically restoring.
+                restorationBlockDepth <- WS.RestorationBlockDepth . view difficultyL <$> DB.getTipHeader
+
+                -- Mark this wallet as officially in restore. As soon as we will pass the point where
+                -- the 'RestorationBlockDepth' is greater than the current store one, we would flip the
+                -- state of this wallet to a "normal" sync, and the two paths will be reunited once for all.
+                setWalletRestorationSyncTip db walletId restorationBlockDepth (headerHash genesisBlock)
+
+                -- Once we have a consistent update of the model, we submit the request to the worker.
+                submitSyncRequest (newRestoreRequest credentials restorationBlockDepth)
+
+-- | Restores the wallet balance by looking at the global Utxo and trying to decrypt
+-- each unspent output address. If we get a match, it means it belongs to us.
+restoreWalletBalance :: ( WalletDbReader ctx m
+                        , HasLoggerName m
+                        , CanLog m
+                        , MonadDBRead m
+                        , MonadUnliftIO m
+                        ) => WalletDB -> WalletDecrCredentials -> m ()
+restoreWalletBalance db credentials = do
+    utxo <- filterUtxo walletUtxo
+    -- FIXME(adn): Quite inefficient for now.
+    -- NOTE: Due to the fact that in the current implementation of the wallet the balance
+    -- (as returned in a 'CWallet') is computed by looking at the coins on each account of
+    -- the wallet, we also need to add these addresses to the wallet for the balance
+    -- computation to consider them.
+    forM_ (M.elems utxo) $ \txOutAux -> do
+        whenJust (decryptAddress credentials . getAddress $ txOutAux) (WS.addWAddress db)
```

Comment:
Is this not a total duplication of what we do in `filterUtxo walletUtxo` on line 76? We've already filtered to those outputs where we can decrypt the address.

Reply:
At this stage, it's unfortunately needed duplication. The reason is the following: `addWAddress` has type `addWAddress :: WAddressMeta -> Update ()`, but alas `WAddressMeta` can be obtained only by decrypting the address within a `txOutAux`, as `decryptAddress` has type
`decryptAddress :: WalletDecrCredentials -> Address -> Maybe WAddressMeta`.

In reality the morally correct way to implement that would be to have something like a `filterMap` variant of my `filterUtxo` function, which could filter and apply a transformation on the matched predicate. I think I even attempted something similar but it wasn't easy to do in the current implementation of the DB.UTXO module. Will give another spin, and if I fail, I suspect we might need to stick with the status quo 😉